probe/endpoint/procspy/proc_linux.go

```diff
@@ -68,51 +86,80 @@ func getNetNamespacePathSuffix() string {
 	return netNamespacePathSuffix
 }
 
-// walkNamespacePid does the work of walkProcPid for a single namespace
-func walkNamespacePid(buf *bytes.Buffer, sockets map[uint64]*Proc, namespaceProcs []*process.Process) {
+// Read the connections for a group of processes living in the same namespace,
+// which are found (identically) in /proc/PID/net/tcp{,6} for any of the
+// processes.
+func readProcessConnections(buf *bytes.Buffer, namespaceProcs []*process.Process) (bool, error) {
+	var (
+		errRead  error
+		errRead6 error
+		read     int64
+		read6    int64
+	)
 
-	// Read the connections for the namespace, which are found (identically) in
-	// /proc/PID/net/tcp{,6} for any of the processes in the namespace.
-	var tcpSuccess bool
 	for _, p := range namespaceProcs {
 		dirName := strconv.Itoa(p.PID)
 
-		read, errRead := readFile(filepath.Join(procRoot, dirName, "/net/tcp"), buf)
-		read6, errRead6 := readFile(filepath.Join(procRoot, dirName, "/net/tcp6"), buf)
+		read, errRead = readFile(filepath.Join(procRoot, dirName, "/net/tcp"), buf)
+		read6, errRead6 = readFile(filepath.Join(procRoot, dirName, "/net/tcp6"), buf)
 
 		if errRead != nil || errRead6 != nil {
 			// try next process
 			continue
 		}
+		return read+read6 > 0, nil
+	}
 
-		if read+read6 == 0 {
-			// No connections, don't bother reading /fd/*
-			return
-		}
-
-		tcpSuccess = true
-		break
+	// would be cool to have an or operation between errors
+	if errRead != nil {
+		return false, errRead
 	}
+	if errRead6 != nil {
+		return false, errRead6
+	}
+
+	return false, nil
 
-	if !tcpSuccess {
-		// There's no point in reading /fd/*
-		return
+}
+
+// walkNamespace does the work of walk for a single namespace
+func (w pidWalker) walkNamespace(buf *bytes.Buffer, sockets map[uint64]*Proc, namespaceProcs []*process.Process) error {
+
+	if found, err := readProcessConnections(buf, namespaceProcs); err != nil || !found {
+		return err
 	}
 
-	// Get the sockets for all the processes in the namespace
 	var statT syscall.Stat_t
-	for _, p := range namespaceProcs {
+	var fdBlockCount uint64
+	for i, p := range namespaceProcs {
+
+		// Get the sockets for all the processes in the namespace
 		dirName := strconv.Itoa(p.PID)
 		fdBase := filepath.Join(procRoot, dirName, "fd")
 
+		if fdBlockCount > w.fdBlockSize {
+			// we surpassed the filedescriptor rate limit
+			// TODO: worth selecting on w.stopc?
```

Comment:
@peterbourgon Maybe it's overkill to select on `stopc` here too (for a higher granularity). But, either way, I am curious about how it could be done without conflicting with the `stopc` select block on `(w pidWalker) walk()`. Any ideas?